test/Microsoft.Azure.Devices.Edge.Test.Common/config/EdgeConfigBuilder.cs

```diff
@@ -0,0 +1,160 @@
+        readonly Dictionary<string, IModuleConfigBuilder> moduleBuilders;
+        Option<(string address, string username, string password)> registry;
+
+        public EdgeConfigBuilder(string deviceId)
+        {
+            this.deviceId = deviceId;
+            this.moduleBuilders = new Dictionary<string, IModuleConfigBuilder>();
+            this.registry = Option.None<(string, string, string)>();
+        }
+
+        public void AddRegistryCredentials(string address, string username, string password)
+        {
+            this.registry = Option.Some((address, username, password));
+        }
+
+        public IModuleConfigBuilder AddEdgeAgent(string image = null)
+        {
+            Option<string> imageOption = Option.Maybe(image);
+            var builder = new AgentModuleConfigBuilder(imageOption);
+            this.moduleBuilders.Add(builder.Name, builder);
+            return builder;
+        }
+
+        public IModuleConfigBuilder AddEdgeHub(string image = null)
+        {
+            Option<string> imageOption = Option.Maybe(image);
+            var builder = new HubModuleConfigBuilder(imageOption);
+            this.moduleBuilders.Add(builder.Name, builder);
+            return builder;
+        }
+
+        public IModuleConfigBuilder AddModule(string name, string image)
+        {
+            var builder = new ModuleConfigBuilder(name, image);
+            this.moduleBuilders.Add(builder.Name, builder);
+            return builder;
+        }
+
+        public EdgeConfiguration Build()
+        {
+            // Build all modules *except* edge agent
+            List<ModuleConfiguration> modules = this.moduleBuilders
+                .Where(b => b.Key != "$edgeAgent")
+                .Select(b => b.Value.Build())
+                .ToList();
+
+            // Build edge agent
+            modules.Insert(0, this.BuildEdgeAgent(modules));
+
+            // Compose edge configuration
+            var config = new ConfigurationContent
+            {
+                ModulesContent = new Dictionary<string, IDictionary<string, object>>()
+            };
+
+            var moduleNames = new List<string>();
+
+            foreach (ModuleConfiguration module in modules)
+            {
+                moduleNames.Add(module.Name);
+
+                if (module.DesiredProperties.Count != 0)
+                {
+                    config.ModulesContent[module.Name] = new Dictionary<string, object>
+                    {
+                        ["properties.desired"] = module.DesiredProperties
+                    };
+                }
+            }
+
+            return new EdgeConfiguration(this.deviceId, moduleNames, config);
+        }
+
+        ModuleConfiguration BuildEdgeAgent(IEnumerable<ModuleConfiguration> configs)
+        {
+            if (!this.moduleBuilders.TryGetValue("$edgeAgent", out IModuleConfigBuilder agentBuilder))
+            {
+                agentBuilder = this.AddEdgeAgent();
```

Comment:
Add log?

Reply:
No, it's perfectly valid for a test to make a basic config that doesn't add anything special to edge agent, in which case we'll quietly add the agent to the config with default settings. There's no reason to log that.

Reply:
do we log what image is used?

Reply:
We used to (i.e., we do in IotEdgeQuickstart), but I think that detail may have been lost in this new library. It would be nice to have it; let me check...